proposals/public-suffix.md

```diff
@@ -0,0 +1,555 @@
+| [adblock-plus](https://addons.mozilla.org/en-US/firefox/addon/adblock-plus/) | #1, #3 | create organization/3rd-party request-blocking rules |
+| [duckduckgo-for-firefox](https://addons.mozilla.org/en-US/firefox/addon/duckduckgo-for-firefox/) | #3 | exclude requests from tracking detection unless they are 3rd-party |
+| [violentmonkey](https://addons.mozilla.org/en-US/firefox/addon/violentmonkey/) | #1 | run scripts automatically for requests having a specific organization name plus any suffix |
+
+## Specification
+
+### Schema
+
+A new API `publicSuffix` is added as follows:
+
+```ts
+namespace publicSuffix {
+  //
+  // Object containing both registrable domains (i.e. ICANN/unknown, Private)
+  // for a domain.
+  //
+  interface RegistrableDomains {
+    // The original domain whose registrable domain we want
+    domain: string,
+    // The ICANN-suffixed or unknown-suffixed registrable domain.
+    // Null if an error occurred.
+    base?: string,
+    // The Private-suffixed registrable domain.
+    // Null if an error occurred, or if the domain has no matching Private suffix.
+    private?: string,
+    // Error thrown during lookup, if any.
+    error?: Error,
+  }
+
+  //
+  // Options that may be passed to getRegistrableDomain() to control its behaviour.
+  //
+  interface RegistrableDomainOptions {
+    // If true, exclude private (non-ICANN) suffixes from the lookup algorithm
+    excludePrivateSuffixes: boolean,
+    // If true, use Punycode instead of Unicode when returning the registrable domain
+    punycode: boolean,
+  }
+
+  //
+  // Gets the longest registrable domain for a specified domain.
+  //
+  // Example:
+  //
+  // let domain = await browser.publicSuffix.getRegistrableDomain("www.example.co.uk");
+  // ==> "example.co.uk"
+  //
+  export function getRegistrableDomain(
+    // The domain name whose registrable domain we want to find
+    domain: string,
+    // Options that control the behaviour of the lookup algorithm
+    options?: RegistrableDomainOptions,
+  )
+  // Resolves to the longest registrable domain of the input domain name
+  : Promise<string>;
+
+  //
+  // Gets both registrable domains (i.e. ICANN/unknown, Private) for each domain name
+  // in an array of domain names.
+  //
+  // Example:
+  //
+  // let domains = await browser.publicSuffix.getRegistrableDomains([
+  //   "foo.bar.wixsite.com",
+  //   "www.example.net",
+  //   "a..b",
+  // ]);
+  // ==> [
+  //   { domain: "foo.bar.wixsite.com", base: "wixsite.com", private: "bar.wixsite.com" },
+  //   { domain: "www.example.net",     base: "example.net",                            },
+  //   { domain: "a..b",                error: "Invalid domain name",                   },
+  // ]
+  //
+  export function getRegistrableDomains(
```

Comment:
Are callers gaining anything from calling this with an array instead of calling the other function multiple times?